python/cuml/ts/arima.py

```diff
@@ -0,0 +1,476 @@
+    p, d, q = order
+
+    if d == 0:
+
+        ll_b, vs = batched_kfilter(np.asfortranarray(y), # numpy
+                                   mu_ar_ma_params_x,
+                                   p, d, q,
+                                   initP_kalman_iterations)
+    elif d == 1:
+
+        y_diff_centered = diffAndCenter(y, p, q, mu_ar_ma_params_x)
+        # print("ydiff:", y_diff_centered)
+        ll_b, vs = batched_kfilter(y_diff_centered, # numpy
+                                   mu_ar_ma_params_x,
+                                   p, d, q,
+                                   initP_kalman_iterations)
+    else:
+        raise NotImplementedError("ARIMA only support d==0,1")
+
+    pynvtx_range_pop()
+    return ll_b, vs
+
+
+def predict_in_sample(model):
+    """Return in-sample prediction on batched series given batched model"""
+
+    p, d, q = model.order[0]
+    x = pack(p, d, q, model.num_batches, model.mu, model.ar_params, model.ma_params)
+    _, vs = run_kalman(model.y, (p, d, q), model.num_batches, x)
+
+    assert_same_d(model.order) # We currently assume the same d for all series
+    _, d, _ = model.order[0]
+
+    if d == 0:
+        y_p = model.y - vs
+    elif d == 1:
+        y_diff = np.diff(model.y, axis=0)
+        # Following statsmodel `predict(typ='levels')`, by adding original
+        # signal back to differenced prediction, we retrive a prediction of
+        # the original signal.
+        predict = (y_diff - vs)
+        y_p = model.y[0:-1, :] + predict
+    else:
+        # d>1
+        raise NotImplementedError("Only support d==0,1")
+
+    # Extend prediction by 1 when d==1
+    if d == 1:
+        # forecast a single value to make prediction length of original signal
+        fc1 = np.zeros(model.num_batches)
+        for i in range(model.num_batches):
+            fc1[i] = fc_single(1, model.order[i], y_diff[:,i],
+                               vs[:,i], model.mu[i],
+                               model.ma_params[i],
+                               model.ar_params[i])
+
+        final_term = model.y[-1, :] + fc1
+
+        # append final term to prediction
+        temp = np.zeros((y_p.shape[0]+1, y_p.shape[1]))
+        temp[:-1, :] = y_p
+        temp[-1, :] = final_term
+        y_p = temp
+
+    model.yp = y_p
+    return y_p
+
+def fc_single(num_steps, order, y_diff, vs, mu, ma_params, ar_params):
+
+    p, _, q = order
+
+    y_ = np.zeros(p+num_steps)
+    vs_ = np.zeros(q+num_steps)
+    if p>0:
+        y_[:p] = y_diff[-p:]
+    if q>0:
+        vs_[:q] = vs[-q:]
+
+    fcast = np.zeros(num_steps)
+
+    for i in range(num_steps):
+        mu_star = mu * (1-ar_params.sum())
+        fcast[i] = mu_star
+        if p > 0:
+            fcast[i] += np.dot(ar_params, y_[i:i+p])
+        if q > 0 and i < q:
+            fcast[i] += np.dot(ma_params, vs_[i:i+q])
+        if p > 0:
+            y_[i+p] = fcast[i]
+
+    return fcast
+
+
+def forecast(model, nsteps: int) -> np.ndarray:
```

Comment:
Same comments as above, part of ARIMAModel and use input_utils, also it should return a GPU vector instead of CPU numpy, no?